xdoc/file_read_text.rst

```diff
@@ -0,0 +1,187 @@
+    # This is slower than the next fgets() example because of string handling
+    #
+    prepare()
+    if open(F, fname):
+      let s = F.fgetsAll
+      cntL = s.countLines
+      cntC = s.len
+      close(F)
+    finalise("fgets all: ")
+      
+    # using fgets() from stdlib of C
+    # 
+    # This shows how to make C library calls
+    # Safety-wise, it is best to use Nim's system lib procs
+    #
+    prepare()
+    if open(F, fname):
+      for s in F.fgetsLine:
+        inc cntL
+        cntC += s.len - 1   # exclude nl char
+      close(F)
+    finalise("fgets iterator: ")
+    
+    # Reading line-by-line using the readLine() iterator from the Nim system lib
+    #
+    # Compare this readLine() with the iterator lines()
+    #
+    prepare()
+    if open(F, fname):
+      while F.readLine(line):
+        inc cntL
+        cntC += line.len
+      close(F)
+    finalise("readLine: ")
+  
+    # using the lines() iterator from the Nim system lib
+    #
+    prepare()
+    if open(F, fname):
+      for line in F.lines:
+        inc cntL
+        cntC += line.len
+      close(F)
+    finalise("lines: ")
+  
+    # using the Nim parseCsv lib
+    #
+    # If you needed to manipulate string portions of each line, 
+    # then this may be a quicker approach, 
+    # because it provides the line already split 
+    # by delimeter (if required)
+    #
+    prepare()
+    var s = newFileStream(fname, fmRead)
+    if s != nil: 
+      var x: CsvParser
+      open(x, s, fname)
+      while x.readRow():
+        inc cntL
+        for z in x.row:
+          cntC += z.len
+      close(x)
+    finalise("parsecsv: ")
+  
+    # using readAll() and splitLines() from Nim system lib
+    #
+    # This would not be appropriate for large files due to
+    # memory usage in reading ALL the file into a string
+    # but for small files works well
+    #
+    prepare()
+    if open(F, fname):
+      let x = F.readAll
+      for line in x.splitLines:
+        inc cntL
+        cntC += line.len
+      close(F)
+      dec cntL
+    finalise("readAll splitLines: ")
+    
+    # using readAll() and splitLines() from Nim system lib
+    #
+    # This is a minor improvement on string handling compared
+    # to the previous example (if you don't need access to the 
+    # full text from the file)
+    #
+    prepare()
+    if open(F, fname):
+      for line in F.readAll().splitLines:
+        inc cntL
+        cntC += line.len
+      close(F)
+      dec cntL
+    finalise("readAll().splitLines: ")
+    
+    # using readAll() and countLines() from Nim system and strutils lib
+    #
+    # This avoids manipulating the file on a per-line basis
+    # and is probably not a practical example for most cases,
+    # but does highlight that you may make performance gains if
+    # you optimise your code for a specific situation
+    #
+    prepare()
+    if open(F, fname):
+      let x = F.readAll
+      cntL = x.countLines
+      cntC = x.len() - (cntL * "\n".len)
```

Comment:
This is wrong. ``countLines`` can deal with any line ending (or should be able to deal with any line ending) and your "\n".len is platform specific, but the file might not use the platform's line endings... This is tricky.